### DISCUSSION TITLE
How does the EdgeDB data model map to the underlying postgres database?

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#1802] [General]
In particular:
1. How is polymorphism mapped?
2. How are links mapped? I'd guess that:
    * 1:1 (single link with constraint exclusive) is a foreign key on the source table with a unique index
    * n:1 (single link) is a foreign key on the source table
    * 1:n (multi link with constraint exclusive) is a foreign key on the target table
    * n:n (multi-link) adds a linker table

**Comment by elprans** [comment]
> 1. How is polymorphism mapped?

Each type gets its own table.  Polymorphic queries use a combination of inline `UNION`s and views over `UNION`s.

> 2. How are links mapped?

`single` links are a column in the source table
`multi` links use link tables

`exclusive` is either a unique index on a source table in the `single` case, or a unique index on the `source` column in the link table.

We don't use foreign keys, because there's no need to: EdgeQL structurally doesn't allow you to write an invalid reference to a table, or leave a dangling reference, but also foreign keys don't work with polymorphic or partitioned tables.

**Reply by CodesInChaos** [reply]
> Each type gets its own table.

How is this compatible with exclusive constraints (without deferred) on abstract types?

> `multi` links use link tables

3. Why use a link table instead of treating it like a single-link in the reverse direction?
4. How do you enforce the `required` constraint on multi-links?
5. Do you support surjective links?

> We don't use foreign keys,

I meant foreign keys in the colloquial sense of "a column containing the primary key of another table". Though that leads to the next question:

6. How do you enforce this, if not through postgres constraints?

**Reply by elprans** [reply]
> How is this compatible with exclusive constraints (without deferred) on abstract types?

Currently, via constraint triggers, but we'll be replacing them with shadow tables soon.

> Why use a link table instead of treating it like a single-link in the reverse direction?

1:n relations are relatively rare to warrant this.  Also, addition or removal of the exclusive constraint would require an expensive data rewrite.

> How do you enforce the `required` constraint on multi-links?

Through a combination of checks in the generated `INSERT/UPDATE` and `on target delete`, but I think it's currently somewhat broken.

> Do you support surjective links?

Not currently, although we can come close enough by implementing `on source delete` policies on links to enforce cascaded deletion when link source is deleted.

> How do you enforce this, if not through postgres constraints?

Referential consistency is inherent in EdgeQL, because there's no way for you to directly populate the foreign keys with values, you must always assign a valid set of objects.  We could add explicit foreign keys + triggers as an extra validation layer in case there is a bug in the query compiler, but that's very expensive and goes in the way of certain migrations.
